core/src/main/java/tech/pegasys/eth2signer/core/service/jsonrpc/FcMessageEncoder.java

```diff
@@ -0,0 +1,108 @@
+    return createFilecoinCid(encodedHashAndCode);
+  }
+
+  private Bytes cborEncode(final FilecoinMessage message) {
+    final byte FILECOIN_MESSAGE_PREFIX = (byte) 138;
+    final CBORFactory cborFactory = new CBORFactory();
+    final ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+    try {
+      // need to ensure _cfgMinimalInts is set in cborgen ... no idea how.
+      final CBORGenerator gen = cborFactory.createGenerator(outputStream);
+      outputStream.write(FILECOIN_MESSAGE_PREFIX);
+      gen.writeNumber(message.getVersion());
+      gen.writeBinary(FilecoinAddress.decode(message.getTo()).getEncodedBytes().toArrayUnsafe());
+      gen.writeBinary(FilecoinAddress.decode(message.getFrom()).getEncodedBytes().toArrayUnsafe());
+      encodeUint64Value(message.getNonce(), gen);
+      serialiseBigInteger(message.getValue(), gen);
+      gen.writeNumber(message.getGasLimit());
+      serialiseBigInteger(message.getGasFeeCap(), gen);
+      serialiseBigInteger(message.getGasPremium(), gen);
+      encodeUint64Value(message.getMethod(), gen);
+      final Bytes paramBytes = Bytes.fromBase64String(message.getParams());
+      gen.writeBinary(null, paramBytes.toArrayUnsafe(), 0, paramBytes.size());
+      gen.close();
+      return Bytes.wrap(outputStream.toByteArray());
+    } catch (IOException e) {
+      throw new IllegalStateException("Failed to create cbor encoded message");
+    }
+  }
+
+  private void encodeUint64Value(final UInt64 value, final CBORGenerator gen) throws IOException {
+    if (value.fitsInt()) {
+      gen.writeNumber(value.intValue());
+    } else if (value.fitsLong()) {
+      gen.writeNumber(value.toLong());
+    } else {
+      final BigInteger bi = value.toBigInteger();
+      gen.writeRaw((byte) (PREFIX_TYPE_INT_POS + SUFFIX_UINT64_ELEMENTS));
+      final Bytes bytes = Bytes.wrap(bi.toByteArray());
+      final int numericBytes = bytes.size() - 1;
+      gen.writeBytes(bytes.slice(1, numericBytes).toArrayUnsafe(), 0, numericBytes);
+    }
+  }
+
+  // Roughly corresponds with filecoin:multhash:Encode
+  private Bytes encodeHashWithCode(final Bytes hashBytes) {
+    final BigInteger FC_HASHING_ALGO_CODE = BigInteger.valueOf(0xb220); // aka Blake2b-256
+    return Bytes.concatenate(
+        ByteUtils.putUVariant(FC_HASHING_ALGO_CODE),
+        ByteUtils.putUVariant(BigInteger.valueOf(hashBytes.size())),
+        hashBytes);
+  }
+
+  // Roughly corresponds to NewCidV1
+  private Bytes createFilecoinCid(final Bytes encodedHashAndCode) {
+    final byte CID_VERSION = (byte) 1;
+    final byte DagCBOR_CODEC_ID = (byte) 113;
+    return Bytes.concatenate(
+        Bytes.wrap(new byte[] {CID_VERSION}),
+        Bytes.wrap(new byte[] {DagCBOR_CODEC_ID}),
+        encodedHashAndCode);
+  }
+
+  private void serialiseBigInteger(final BigInteger value, final CBORGenerator gen)
+      throws IOException {
+    if (value.equals(BigInteger.ZERO)) {
```

Comment:
BigInteger should use compareTo instead of equals.

Reply:
That's true for `BigDecimal` but why for `BigInteger`?

Reply:
This should be Ok for BigInteger (which explicitly states that equals is numeric equivalence) - BigDecimal may be an issue - but isn't being used here.

Reply:
I take my statement back :), my mind crossed it with BigDecimal